src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -2548,6 +2647,326 @@ public final void addBatch() throws SQLServerException {
         loggerExternal.exiting(getClassNameLogging(), "executeLargeBatch", updateCounts);
         return updateCounts;
     }
+    
+    
+    private String parseUserSQLForTableNameDW(boolean hasInsertBeenFound, boolean hasIntoBeenFound) {
+        // As far as finding the table name goes, There are two cases:
+        // Insert into <tableName> and Insert <tableName>
+        // And there could be in-line comments (with /* and */) in between.
+        // This method assumes the localUserSQL string starts with "insert".
+        localUserSQL = localUserSQL.trim();
+        if (checkAndRemoveComments()) {
+            return parseUserSQLForTableNameDW(hasInsertBeenFound, hasIntoBeenFound);
+        }
+        
+        if (localUserSQL.substring(0, 6).equalsIgnoreCase("insert") && !hasInsertBeenFound) {
+            localUserSQL = localUserSQL.substring(6);
+            return parseUserSQLForTableNameDW(true, hasIntoBeenFound);
+        }
+        
+        if (localUserSQL.substring(0, 4).equalsIgnoreCase("into") && !hasIntoBeenFound) {
+            // is it really "into"?
+            // if the "into" is followed by a blank space or /*, then yes.
+            if (localUserSQL.charAt(4) == ' ' || localUserSQL.charAt(4) == '\t' ||
+                    (localUserSQL.charAt(4) == '/' && localUserSQL.charAt(5) == '*')) {
+                localUserSQL = localUserSQL.substring(4);
+                return parseUserSQLForTableNameDW(hasInsertBeenFound, true);
+            }
+            
+            // otherwise, we found the token that either contains the databasename.tablename or tablename.
+            // Recursively handle this, but into has been found. (or rather, it's absent in the query - the "into" keyword is optional)
+            return parseUserSQLForTableNameDW(hasInsertBeenFound, true);
+        }
+        
+        // At this point, the next token has to be the table name.
+        // It could be encapsulated in [], "", or have a database name preceding the table name.
+        // If it's encapsulated in [] or "", we need be more careful with parsing as anything could go into []/"".
+        // For ] or ", they can be escaped by ]] or "", watch out for this too.
+        if (localUserSQL.substring(0, 1).equalsIgnoreCase("[")) {
+            int tempint = localUserSQL.indexOf("]", 1);
+            
+            // keep checking if it's escaped
+            while (localUserSQL.charAt(tempint + 1) == ']') {
+                localUserSQL = localUserSQL.substring(0, tempint) + localUserSQL.substring(tempint + 1);
+                tempint = localUserSQL.indexOf("]", tempint + 1);
+            }
+            
+            // we've found a ] that is actually trying to close the square bracket.
+            // If it's followed by a dot, then it's a database.
+            // Otherwise, it's the table.
+            if (localUserSQL.charAt(tempint + 1) == '.') {
```

Comment:
This is a valid query, notice the whitespaces  `insert into [master]     .               dbo            .         tablename values(1)`, which would probably fail here.

Reply:
Yes. I had to change some of the logic for parsing table names, including adding a couple more flags. Thanks for the catch!